src/Roslyn.Suppressions/Roslyn.Suppressions.UnitTests/SuppressedIssuesFileStorageTests.cs

```diff
@@ -0,0 +1,298 @@
+            act.Should().ThrowExactly<ArgumentException>().And.ParamName.Should().Be("sonarProjectKey");
+        }
+
+        [TestMethod]
+        public void Update_ProjectKeyHasInvalidChars_InvalidCharsReplaced()
+        {
+            SonarQubeIssue issue = CreateIssue("issueKey");
+            var issues = new List<SonarQubeIssue> { issue };
+
+            var logger = new TestLogger();
+
+            var file = new Mock<IFile>();
+            Mock<IFileSystem> fileSystem = CreateFileSystem(file);
+
+            var fileStorage = new SuppressedIssuesFileStorage(logger, fileSystem.Object);
+            fileStorage.Update("project:Key", issues);
+
+            file.Verify(f => f.WriteAllText(GetFilePath("project_Key"), JsonConvert.SerializeObject(issues)), Times.Once);
+            logger.AssertNoOutputMessages();
+        }
+
+        [DataRow(null)]
+        [DataRow("")]
+        [DataRow("  ")]
+        [TestMethod]
+        public void Get_ProjectKeyIsEmpty_ExceptionThrown(string projectKey)
+        {
+            var fileStorage = CreateTestSubject();
+
+            Action act = () => fileStorage.Get(projectKey);
+            act.Should().ThrowExactly<ArgumentException>().And.ParamName.Should().Be("sonarProjectKey");
+        }
+
+        [TestMethod]
+        public void Get_ProjectKeyHasInvalidChars_InvalidCharsReplaced()
+        {
+            SonarQubeIssue issue = CreateIssue("issueKey");
+            var issues = new List<SonarQubeIssue> { issue };
+
+            var logger = new TestLogger();
+
+            var file = CreateFileForGet("project_Key", issues);
+            Mock<IFileSystem> fileSystem = CreateFileSystem(file);
+
+            var fileStorage = CreateTestSubject(logger, fileSystem.Object);
+
+            var issuesGotten = fileStorage.Get("project:Key").ToList();
+
+            file.Verify(f => f.ReadAllText(GetFilePath("project_Key")), Times.Once);
+            logger.AssertNoOutputMessages();
+
+            issuesGotten.Count.Should().Be(1);
+            issuesGotten[0].IssueKey.Should().Be(issue.IssueKey);
+        }
+
+        [TestMethod]
+        public void Update_ErrorOccuredWhenWritingFile_ErrorIsLogged()
+        {
+            SonarQubeIssue issue = CreateIssue("issueKey");
+            var issues = new List<SonarQubeIssue> { issue };
+
+            var logger = new TestLogger();
+
+            var file = new Mock<IFile>();
+            file.Setup(f => f.WriteAllText(It.IsAny<string>(), It.IsAny<string>())).Throws(new Exception("Test Exception"));
+            Mock<IFileSystem> fileSystem = CreateFileSystem(file);
+
+            var fileStorage = CreateTestSubject(logger, fileSystem.Object);
+
+            fileStorage.Update("projectKey", issues);
+
+            logger.AssertOutputStrings("[Roslyn Suppressions] Error writing settings for project projectKey. Issues suppressed on the server may not be suppressed in the IDE. Error: Test Exception");
```

Comment:
FYI if a logged message has parameters, we don't tend to check the fully-constructed strings - we would normally just check the expected parameters are present using one of the `TestLogger` assertion methods like `AssertPartialOutputStrings`.
There's no particular problem with checking for the full string, it's just not something we tend to do.